pyccel/parser/syntax/openmp.py

```diff
@@ -11,396 +11,962 @@
 from textx.metamodel import metamodel_from_file
 
 from pyccel.parser.syntax.basic import BasicStmt
-from pyccel.ast.core import OMP_For_Loop, OMP_Parallel_Construct, OMP_Single_Construct, Omp_End_Clause
+
+from pyccel.ast.omp import (OmpAnnotatedComment, OMP_For_Loop, OMP_Parallel_Construct,
+                            OMP_Single_Construct, Omp_End_Clause, OMP_Critical_Construct,
+                            OMP_Master_Construct, OMP_Masked_Construct, OMP_Task_Construct,
+                            OMP_Cancel_Construct, OMP_Target_Construct, OMP_Teams_Construct,
+                            OMP_Sections_Construct, OMP_Section_Construct)
 
 DEBUG = False
 
+class OmpConstruct(BasicStmt):
+    """Class representing all OpenMP constructs."""
+    def __init__(self, omp_type, vclauses, **kwargs):
+        name     = kwargs.pop('name', None)
+        clauses  = kwargs.pop('clauses', None)
+        combined = kwargs.pop('combined', None)
+
+        _valid_clauses = vclauses
+
+        com = None
+        if combined:
+            if 'for' in combined.expr:
+                _valid_clauses += _valid_loop_clauses
+            if 'simd' in combined.expr:
+                _valid_clauses += _valid_simd_clauses
+            if 'taskloop' in combined.expr:
+                _valid_clauses += _valid_taskloop_clauses
+            if 'masked' in combined.expr:
+                _valid_clauses += (OmpFilter,)
+            if 'sections' in combined.expr:
+                _valid_clauses += _valid_sections_clauses
+            if 'distribute' in combined.expr:
+                _valid_clauses += _valid_Distribute_clauses
+            if 'parallel' in combined.expr:
+                _valid_clauses += _valid_parallel_clauses
+            if 'teams' in combined.expr:
+                _valid_clauses += _valid_teams_clauses
+            com = combined.expr
+
+        txt = ''
+        if name:
+            txt += name
+        if clauses:
+            txt += check_get_clauses(self, _valid_clauses, clauses, combined)
+
+        if combined:
+            self._expr = omp_type(txt, com)
+        else:
+            self._expr = omp_type(txt)
+
+        super().__init__(**kwargs)
+
+    @property
+    def expr(self):
+        if DEBUG:
+            print("> {}: expr".format(type(self).__name__))
+
+        return self._expr
+
+def check_get_clauses(name, valid_clauses, clauses, combined = None):
+    """
+    Function to check if the clauses are correct for a given Construct.
+    """
+    txt = ''
+    for clause in clauses:
+        if isinstance(clause, valid_clauses):
+            if isinstance(clause, OmpCopyin) and isinstance(combined, OmpTargetParallel):
```

Comment:
can these checks be in one if block since the next statement is just an else?

Reply:
I agree, it would be neater to avoid the duplicate code with:
```
if isinstance(clause, valid_clauses) and \
   not (isinstance(clause, OmpCopyin) and isinstance(combined, OmpTargetParallel)):
```